core/pass.cpp

```diff
@@ -282,70 +282,50 @@ void CompilerPass::visit(Unary *ast)
+     auto *ast = static_cast<astClass *>(var); \
+     visit(ast); \
+   } break
+
 void CompilerPass::visitExpr(AST *&ast_)
 {
-    if (auto *ast = dynamic_cast<Apply *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<ApplyBrace *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Array *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<ArrayComprehension *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Assert *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Binary *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<BuiltinFunction *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Conditional *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Dollar *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Error *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Function *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Import *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Importstr *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<InSuper *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Index *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Local *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<LiteralBoolean *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<LiteralNumber *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<LiteralString *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<LiteralNull *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Object *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<DesugaredObject *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<ObjectComprehension *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<ObjectComprehensionSimple *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Parens *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Self *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<SuperIndex *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Unary *>(ast_)) {
-        visit(ast);
-    } else if (auto *ast = dynamic_cast<Var *>(ast_)) {
-        visit(ast);
-
-    } else {
-        std::cerr << "INTERNAL ERROR: Unknown AST: " << ast_ << std::endl;
-        std::abort();
+    switch(ast_->type) {
+        VISIT(ast_, AST_APPLY, Apply);
+        VISIT(ast_, AST_APPLY_BRACE, ApplyBrace);
+        VISIT(ast_, AST_ARRAY, Array);
+        VISIT(ast_, AST_ARRAY_COMPREHENSION, ArrayComprehension);
+        // VISIT(ast_, AST_ARRAY_COMPREHENSION, ArrayComprehensionSimple);
```

Comment:
What's happening here?

Reply:
I think all these passes run before desugaring, so there should not be any ArrayComprehensionSimple.  That could change in future of course.